dcos/auth.py

```diff
@@ -0,0 +1,308 @@
+
+    creds = {"uid": username, "password": password}
+    return _get_dcostoken_by_post_with_creds(dcos_url, creds)
+
+
+def dcos_uid_password_auth(dcos_url, username=None, password=None):
+    """
+    Authenticate user using DC/OS uid password auth
+
+    Raises exception if authentication fails.
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :param username: username to auth with
+    :type username: str
+    :param password: password to auth with
+    :type password: str
+    :rtype: None
+    """
+
+    dcos_token = _get_dcostoken_by_dcos_uid_password_auth(
+            dcos_url, username, password)
+    if not dcos_token:
+        raise DCOSException("Authentication failed")
+    else:
+        return
+
+
+def oidc_implicit_flow_auth(dcos_url):
+    """
+    Authenticate user using OIDC implict flow
+
+    Raises exception if authentication fails.
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :rtype: None
+    """
+
+    dcos_token = _get_dcostoken_by_oidc_implicit_flow(dcos_url)
+    if not dcos_token:
+        raise DCOSException("Authentication failed")
+    else:
+        return
+
+
+def _get_dcostoken_by_oidc_implicit_flow(dcos_url):
+    """
+    Get DC/OS Authentication Token by OIDC implicit flow
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :returns: DC/OS Authentication Token
+    :rtype: str
+    """
+
+    oauth_login = 'login?redirect_uri=urn:ietf:wg:oauth:2.0:oob'
+    url = urllib.parse.urljoin(dcos_url, oauth_login)
+    oidc_token = _prompt_user_for_token(url, "OpenID Connect ID")
+    creds = {"token": oidc_token}
+    return _get_dcostoken_by_post_with_creds(dcos_url, creds)
+
+
+def servicecred_auth(dcos_url, username, key_path):
+    """
+    Get DC/OS Authentication token by browser prompt
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :param username: username user for authentication
+    :type username: str
+    :param key_path: path to service key
+    :param key_path: str
+    :rtype: None
+    """
+
+    creds = {
+        'uid': username,
+        'token': jwt.encode(
+            {
+                'exp': int(time.time()+5*60),
+                'uid': username
+            },
+            util.read_file(key_path),
+            algorithm='RS256')
+        .decode('ascii')
+    }
+
+    dcos_token = _get_dcostoken_by_post_with_creds(dcos_url, creds)
+    if not dcos_token:
+        raise DCOSException("Authentication failed")
+    else:
+        return
+
+
+def browser_prompt_auth(dcos_url, provider_info):
+    """
+    Get DC/OS Authentication token by browser prompt
+
+    :param dcos_url: url to cluster
+    :type dcos_url: str
+    :param provider_info: info about provider to auth with
+    :param provider_info: str
+    :rtype: None
+    """
+
+    start_flow_url = provider_info["client"]["start_flow_url"]
+    if not urlparse(start_flow_url).netloc:
+        start_flow_url = urljoin(dcos_url, start_flow_url)
```

Comment:
Again, I am strongly in favor of controlling the trailing slash behavior of `dcos_url` elsewhere plus using simple string concatenation here.